### PR TITLE
planner/cascades[WIP]: add transformation rule TransformJoinCondToSel

### PR DESCRIPTION
**Pull request by SeaRise**
### What problem does this PR solve?
sql `select * from t1 join t2 on t1.a = t2.a and t1.a > 2` In cascades planner will fail.
Because of join's rightCond has not be pushed down.
Add transformation rule `TransformJoinCondToSel` to push down join's cond.
And fix some problems in `PushSelDownJoin`.

### Check List

Tests

- Unit test
- Integration test

### CODE REVIEW

**Comment by francis0407** [comment]
Sorry, I cannot understand the goal of this rule.
There are bugs in `PushSelDownJoin` because `DeriveOtherConditions` calls `p.children`. But I think it can push down right conditions by `rightGroup = buildChildSelectionGroup(sel, rightCond, rightGroup)`.

**Comment by SeaRise** [comment]
@francis0407 
`PushSelDownJoin` match `sel->join`.
But for sql `select * from t1 join t2 on t1.a = t2.a and t1.a > 2`, there are no `Sel` behind `join`.

**Comment by francis0407** [comment]
> @francis0407
> `PushSelDownJoin` match `sel->join`.
> But for sql `select * from t1 join t2 on t1.a = t2.a and t1.a > 2`, there are no `Sel` behind `join`.

Good Catch, I get it now.

File: planner/cascades/transformation_rules.go [hunk 1]
```diff
@@ -953,10 +925,68 @@ func (r *PushSelDownJoin) OnTransform(old *memo.ExprIter) (newExprs []*memo.Grou
...
+	leftCond, rightCond, remainCond, dual := r.predicatePushDownJoin(sctx, sel.Conditions, newJoin, leftGroup.Prop.Schema, rightGroup.Prop.Schema)
+	if dual != nil {
+		return []*memo.GroupExpr{memo.NewGroupExpr(dual)}, true, true, nil
+	}
+
 	// TODO: Update EqualConditions like what we have done in the method join.updateEQCond() before.
-	leftGroup = buildChildSelectionGroup(sel, leftCond, leftGroup)
-	rightGroup = buildChildSelectionGroup(sel, rightCond, rightGroup)
-	newJoinExpr := memo.NewGroupExpr(join)
+	leftGroup = buildChildSelectionGroup(sel.SCtx(), sel.SelectBlockOffset(), leftCond, leftGroup)
```
Comment (francis0407) [inline planner/cascades/transformation_rules.go]: We need to consider the join type here. For example:
`select * from t1 left join t2 on t1.a > 10`, the condition `t1.a > 10` cannot be pushed down.

Comment (SeaRise) [reply planner/cascades/transformation_rules.go]: if is right join, `len(leftCond) == 0`?.I'm not sure.I will add a test for it.
Ok, I will wait for  #15426.May be We should `TransformApplyCondToSel` too.

**Pull request closed without merging**